src/liballoc/collections/btree/node.rs

```diff
@@ -460,19 +460,19 @@ impl<'a, K, V, Type> NodeRef<marker::Mut<'a>, K, V, Type> {
 }
 
 impl<'a, K: 'a, V: 'a, Type> NodeRef<marker::Immut<'a>, K, V, Type> {
-    /// SAFETY: the caller must ensure that the node has more than `idx` elements.
+    /// SAFETY: the caller must ensure that the node has initialized more than `idx` elements.
     unsafe fn into_key(self, idx: usize) -> &'a K {
         debug_assert!(idx < self.len());
-        unsafe { &MaybeUninit::get_ref(&(*self.as_leaf()).keys[idx]) }
+        unsafe { &MaybeUninit::get_ref(&(*self.as_leaf()).keys.get_unchecked(idx)) }
```

Comment:
Now this is surprising that this works... `get_unchecked` needs a reference to the entire slice, and only then performs indexing. Really what we want here is unchecked indexing on the raw slice.

I think the only way to do this currently is to do a cast of `&raw const (*self.as_leaf()).keys` to `*const K`, then do `add(idx)`.

Reply:
I'm lost, but I'll try to feel my way out again. Beware that (what I thought is) this change from checked to unchecked indexing is for the immutable case which wasn't a problem. When I replace `[]` with `.get_unchecked_mut` in the third function `into_key_val_mut`,  the one where it matters for Miri and so that I had forgotten about already, the compiler says the thing resulting from `.get_unchecked_mut`  is a temporary that it can't return. Even adding enough parenthesis to hide my lack of knowledge of operator precedence. Though I can get the stuff compiled with `.get_unchecked_mut` but probably it's not the executing the one from slice because then Miri protests again on the new test case. Again, I confirm it does not protest with checked indexing, or whatever `[]` really is.

Also, surely the first `&` is smoke and mirrors? It compiles and works the same without. `get_ref` returns a reference already.

Reply:
Your old code took a reference to `(*self.as_leaf()).keys[idx]`. That is just a reference to a single element, so all is good -- no aliasing with references to other elements.

Your new code creates a reference to `(*self.as_leaf()).keys`. This *does* always with other elements, hence I think we should not do it.

> for the immutable case which wasn't a problem

Ah I guess that explains why we are not seeing Miri failures.

Reply:
So you're saying that (in C++ terms) operator[] is quite different from `get_unchecked` with a bouncer posted at the front door? Perhaps operator[] is not even the same as plain `get`? That's... unexpected.

>  that explains why we are not seeing Miri failures.

Because it's perfectly fine to refer to read-only slices all over the place... So why was I messing about in that code? I guess it was relevant when that branch started. Now it's still nice to keep passing the index for symmetry with the mutable API, but maybe revert these two implementations to MaybeUnit slice code?

Reply:
Yes. `[]` is a MIR built-in, which is very different from a method call.

I appreciate that from a surface-level perspective, that is rather surprising. Unfortunately I have no idea how to make this less surprising without breaking even more things.^^

> Now it's still nice to keep passing the index for symmetry with the mutable API

Yeah, I think it makes a lot of sense to just generally avoid creating these slices in the first place, just to avoid unnecessary aliasi8ng assumptions.

Reply:
I've simplified the immutable implementations on MaybeUnit again.

Reply:
I think the simplest approach would be to use the raw slice indexing I am adding in https://github.com/rust-lang/rust/pull/73986 -- maybe we can do that in a follow-up PR.

Reply:
Definitely. I noticed that `get_unchecked` is used throughout to access an individual edge from the edges array, and `as_ptr` instead of `MaybeUninit::as_ptr`, so that's the pattern I copied this time, even though I dislike it (more so because they define `as_ptr` as a local method too).